Can't install geopyspark in a python virtualenv environment

Repro case:

- Make a [virtualenv](http://cheng.logdown.com/posts/2016/08/12/install-python3-and-virtualenv-on-mac):
```bash
mkvirtualenv --python=/usr/local/bin/python3 spark
```
- add the SPARK_HOME to virtualenv:
`vim ~/.envs/spark/bin/postactivate`
```bash
export SPARK_HOME=/usr/local/spark
```
- Activate virtualenv and install pygeospark
```bash
workon spark
git clone https://github.com/locationtech-labs/geopyspark.git
cd geopyspark
make install
```

You will get the following error message:

I get the following error:
```bash
python3 setup.py install --user --force --prefix=
running install
Checking .pth file support in /Users/rburhum/.local/lib/python3.6/site-packages/
/Users/rburhum/.virtualenvs/spark/bin/python3 -E -c pass
TEST FAILED: /Users/rburhum/.local/lib/python3.6/site-packages/ does NOT support .pth files
error: bad install directory or PYTHONPATH

You are attempting to install a package to a directory that is not
on PYTHONPATH and which Python does not read ".pth" files from.  The
installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /Users/rburhum/.local/lib/python3.6/site-packages/

and your PYTHONPATH environment variable currently contains:

    '/Users/rburhum/.virtualenvs/spark/lib/python3.6/site-packages/'

Here are some of your options for correcting the problem:

* You can choose a different installation directory, i.e., one that is
  on PYTHONPATH or supports .pth files

* You can add the installation directory to the PYTHONPATH environment
  variable.  (It must then also be on PYTHONPATH whenever you run
  Python and want to use the package(s) you are installing.)

* You can set up the installation directory to support ".pth" files by
  using one of the approaches described here:

  https://setuptools.readthedocs.io/en/latest/easy_install.html#custom-installation-locations


Please make the appropriate changes for your system and try again.
make: *** [install] Error 1
```

If I edit the Makefile like so:
```bash
rburhum@peru:~/geopyspark$ git diff
diff --git a/Makefile b/Makefile
index 6a884a8..fbc66f3 100644
--- a/Makefile
+++ b/Makefile
@@ -10,7 +10,7 @@ WHEEL := dist/${WHEELNAME}
 rwildcard=$(foreach d,$(wildcard $1*),$(call rwildcard,$d/,$2) $(filter $(subst *,%,$2),$d))

 install: ${DIST-ASSEMBLY} ${WHEEL}
-       ${PYTHON} setup.py install --user --force --prefix=
+       ${PYTHON} setup.py install

 ${DIST-ASSEMBLY}: ${BUILD-ASSEMBLY}
        cp -f ${BUILD-ASSEMBLY} ${DIST-ASSEMBLY}
rburhum@peru:~/geopyspark$
```
and to a `make install` it works fine.

--- a/Makefile
+++ b/Makefile
@@ -28,6 +28,8 @@ ${WHEEL}: ${DIST-ASSEMBLY} $(call rwildcard, geopyspark, *.py) setup.py
 
 wheel: ${WHEEL}
 
+build: ${DIST-ASSEMBLY}
+
 pyspark: ${DIST-ASSEMBLY}
 	pyspark --jars ${DIST-ASSEMBLY} \
 		--conf spark.serializer=org.apache.spark.serializer.KryoSerializer \

--- a/Makefile
+++ b/Makefile
@@ -14,8 +14,8 @@ install: ${DIST-ASSEMBLY} ${WHEEL}
 	${PYTHON} setup.py install --user --force --prefix=
 
 virtual-install: ${DIST-ASSEMBLY}
-	pip install -r requirements.txt --force --prefix=${VIRTUAL_ENV} && \
-	${PYTHON} setup.py install --force --prefix=${VIRTUAL_ENV}
+	pip install -r requirements.txt --force && \
+	${PYTHON} setup.py install --force
 
 ${DIST-ASSEMBLY}: ${BUILD-ASSEMBLY}
 	cp -f ${BUILD-ASSEMBLY} ${DIST-ASSEMBLY}

--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,7 @@ install: ${DIST-ASSEMBLY} ${WHEEL}
 	${PYTHON} setup.py install --user --force --prefix=
 
 virtual-install: ${DIST-ASSEMBLY}
+	pip install -r requirements.txt --force --prefix=${VIRTUAL_ENV} && \
 	${PYTHON} setup.py install --force --prefix=${VIRTUAL_ENV}
 
 ${DIST-ASSEMBLY}: ${BUILD-ASSEMBLY}

--- a/Makefile
+++ b/Makefile
@@ -13,6 +13,9 @@ rwildcard=$(foreach d,$(wildcard $1*),$(call rwildcard,$d/,$2) $(filter $(subst 
 install: ${DIST-ASSEMBLY} ${WHEEL}
 	${PYTHON} setup.py install --user --force --prefix=
 
+virtual-install: ${DIST-ASSEMBLY}
+	${PYTHON} setup.py install --force --prefix=${VIRTUAL_ENV}
+
 ${DIST-ASSEMBLY}: ${BUILD-ASSEMBLY}
 	cp -f ${BUILD-ASSEMBLY} ${DIST-ASSEMBLY}
 

--- a/README.rst
+++ b/README.rst
@@ -25,8 +25,8 @@ know what you're working on!) you contact us at:
 As you may have noticed from the above links, those are links to the GeoTrellis
 gitter channel and mailing list. This is because this project is currently an
 offshoot of GeoTrellis, and we will be using their mailing list and gitter
-channel as a means of contact. However, we will form our own if there is
-a need for it.
+channel as a means of contact. However, we will form our own if there is a need
+for it.
 
 Setup
 ------
@@ -43,12 +43,12 @@ Python       3.3 - 3.5
 Hadoop       >=2.0.1
 ============ ============
 
-Java 8 and Scala 2.11 are needed for GeoPySpark to work; as they are required
-by GeoTrellis. In addition, Spark needs to be installed and configured with the
+Java 8 and Scala 2.11 are needed for GeoPySpark to work; as they are required by
+GeoTrellis. In addition, Spark needs to be installed and configured with the
 environment variable, ``SPARK_HOME`` set.
 
-You can test to see if Spark is installed properly by running the following in the
-terminal:
+You can test to see if Spark is installed properly by running the following in
+the terminal:
 
 .. code:: console
 
@@ -64,6 +64,9 @@ How to Install
 Before installing, check the above table to make sure that the
 requirements are met.
 
+Installing From Pip
+~~~~~~~~~~~~~~~~~~~~
+
 To install via ``pip`` open the terminal and run the following:
 
 .. code:: console
@@ -75,39 +78,8 @@ Where the first command installs the python code from PyPi and the second
 downloads the backend, jar file. If no path is given when downloading the jar,
 then it will be downloaded to wherever GeoPySpark was installed at.
 
-If you would rather install from source, you can do so by running the following
-in the terminal:
-
-.. code:: console
-
-   git clone https://github.com/locationtech-labs/geopyspark.git
-   cd geopyspark
-   make install
-
-This will assemble the backend-end ``jar`` that contains the Scala code,
-move it to the ``jars`` module, and then runs the ``setup.py`` script.
-
-Note:
-  If you have somehow altered the global behavior of ``sbt`` this install may
-  not work correctly.
-
-A third option is to install GeoPySpark in a virtual environment. To do this,
-enter the environment and run the following:
-
-.. code:: console
-
-   git clone https://github.com/locationtech-labs/geopyspark.git
-   cd geopyspark
-   export PYTHONPATH=$VIRTUAL_ENV/lib/<your python version>/site-packages
-   make virtual-install
-
-Replace ``<your python version`` with whatever Python version
-``virtualenvwrapper`` is set to. Installation in a virtual environment can be
-a bit weird with GeoPySpark. This is why you need to export the
-``PYTHONPATH`` before installing to ensure that it performs correctly.
-
 What's With That Weird Pip Install?
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+====================================
 
 "What's with that weird pip install?", you may be asking yourself. The reason
 for its unusualness is due to how GeoPySpark functions. Because this library
@@ -120,46 +92,97 @@ method of distribution where the jar must be downloaded in a serperate command
 when using ``pip install``.
 
 Note:
-  Installing from source does not require the seperate download of the jar.
+  Installing from source or for development does not require the seperate
+  download of the jar.
+
+Installing From Source
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If you would rather install from source, clone the GeoPySpark repo and enter it.
+
+.. code:: console
+
+   git clone https://github.com/locationtech-labs/geopyspark.git
+   cd geopyspark
+
+Installing For Users
+=====================
+
+.. code:: console
+
+   make install
+
+This will assemble the backend-end ``jar`` that contains the Scala code,
+move it to the ``jars`` sub-package, and then runs the ``setup.py`` script.
+
+Note:
+  If you have altered the global behavior of ``sbt`` this install may
+  not work the way it was intended.
+
+Installing For Developers
+===========================
+
+.. code:: console
+
+   make build
+   pip install -e .
+
+``make build`` will assemble the back-end ``jar`` and move it the ``jars``
+sub-package. The second command will install GeoPySpark in "editable" mode.
+Meaning any changes to the source files will also appear in your system
+installation.
+
+Installing to a Virtual Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A third option is to install GeoPySpark in a virtual environment. To get things
+started, enter the envrionemnt and run the following:
+
+.. code:: console
+
+   git clone https://github.com/locationtech-labs/geopyspark.git
+   cd geopyspark
+   export PYTHONPATH=$VIRTUAL_ENV/lib/<your python version>/site-packages
+
+Replace ``<your python version`` with whatever Python version
+``virtualenvwrapper`` is set to. Installation in a virtual environment can be
+a bit weird with GeoPySpark. This is why you need to export the
+``PYTHONPATH`` before installing to ensure that it performs correctly.
+
+Installing For Users
+=====================
+
+.. code:: console
+
+   make virtual-install
+
+Installing For Developers
+===========================
+
+.. code:: console
+
+   make build
+   pip install -e .
+
 
 Developing GeoPySpark With GeoNotebook
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 `GeoNotebook <https://github.com/OpenGeoscience/geonotebook>`_ is a Jupyter
-notebook extension that specializes in working with GeoSpatial data. GeoPySpark
-can be used on this notebook; which allows for a more interactive experience
-when using the library. There two setups one could have when using these two
-tools: one for users and the other for developers. Both instances have the same
-begining steps and then diverge later. Therefore, let's take a look at what
-needs to be done first.
+notebook extension that specializes in working with geospatial data. GeoPySpark
+can be used with this notebook; which allows for a more interactive experience
+when using the library. For this section, we will be installing both tools in a
+virtual environment. It is recomended that you start with a new environment
+before following this guide.
 
-.. code:: console
+Because there's already documentation on how to install GeoPySpark in a virtual
+environment, we won't go over it here. As for GeoNotebook, it also has a section
+on `installtion <https://github.com/OpenGeoscience/geonotebook#make-a-virtualenv-install-jupyternotebook-install-geonotebook>`_
+so that will not be covered here either.
 
-   mkdir workspace
-   cd workspace
-   mkvirtualenv workspace
-
-The first thing to do is to create a new directory, ``workspace`` (or whatever
-you want to name it) and then using ``virtualenvwrapper`` to create a new
-``virtualenv`` that will contain both GeoPySpark, GeoNotebook, and their
-dependencies.
-
-Now this is where the installation for a user and developer diverge. To install
-for a user, simply follow the installation instructions for both GeoPySpark and
-GeoNotebook.
-
-For a developer, follow the instructions in the GeoNotebook README on how to
-install for development. To isntall GeoPySpark, clone the latest version from
-Github and then do this command:
-
-.. code:: console
-
-   pip install -e .
-
-This will install GeoPySpark in "editable" mode.
-
-To actaully run the notebooks, go to the folder where they are saved and run
-the command:
+Once you've setup both GeoPySpark and GeoNotebook, all that needs to be done
+is go to where you want to save/have saved your notebooks and execute this
+command:
 
 .. code:: console
 
@@ -167,8 +190,11 @@ the command:
 
 This will open up the jupyter hub and will allow you to work on your notebooks.
 
-One last optional (but very recomended!) step for developers is to place these
-two lines of code at the top of your notebook.
+Side Note For Developers
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An optional (but recomended!) step for developers is to place these
+two lines of code at the top of your notebooks.
 
 .. code:: console
 
@@ -177,38 +203,12 @@ two lines of code at the top of your notebook.
 
 This will make it so that you don't have to leave the notebook for your changes
 to take affect. Rather, you just have to reimport the module and it will be
-updated.
+updated. However, there are a few caveats when using ``autoreload`` that can be
+read `here <http://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html#caveats>`_.
 
-
-Make Targets
-^^^^^^^^^^^^
-
- - **install** - install ``GeoPySpark`` python package locally
- - **wheel** - build python ``GeoPySpark`` wheel for distribution
- - **pyspark** - start pyspark shell with project jars
- - **docker-build** - build docker image for Jupyter with ``GeoPySpark``
- - **clean** - remove the wheel, the backend jar file, and clean the
-   ``geotrellis-backend`` directory
- - **cleaner** - the same as **clean**, but also erase all ``.pyc``
-   files and delete binary artifacts in the ``docker`` directory
-
-Docker Container
-^^^^^^^^^^^^^^^^
-
-To build the docker container, type the following in a terminal:
-
-.. code:: console
-
-   make docker-build
-
-If you encounter problems, typing ``make cleaner`` before typing
-``make docker-build`` could help.
-
-To run the container, type:
-
-.. code:: console
-
-   docker run -it --rm -p 8000:8000 quay.io/geodocker/jupyter-geopyspark:6
+Using ``pip install -e`` in conjunction with ``autoreload`` should cover any
+changes made, though, and will make the development experience much less
+painful.
 
 GeoPySpark Script
 -----------------
@@ -232,6 +232,37 @@ These can be used regardless of installation method. However, if installed
 through ``pip``, then the jar must be downloaded first or these commands
 will not work.
 
+Make Targets
+^^^^^^^^^^^^
+
+ - **install** - install GeoPySpark python package locally
+ - **wheel** - build python GeoPySpark wheel for distribution
+ - **pyspark** - start pyspark shell with project jars
+ - **build** - builds the backend jar and moves it to the jars sub-package
+ - **docker-build** - build docker image for Jupyter with GeoPySpark
+ - **clean** - remove the wheel, the backend jar file, and clean the
+   geotrellis-backend directory
+ - **cleaner** - the same as **clean**, but also erase all .pyc
+   files and delete binary artifacts in the docker directory
+
+Docker Container
+^^^^^^^^^^^^^^^^
+
+To build the docker container, type the following in a terminal:
+
+.. code:: console
+
+   make docker-build
+
+If you encounter problems, typing ``make cleaner`` before typing
+``make docker-build`` could help.
+
+To run the container, type:
+
+.. code:: console
+
+   docker run -it --rm -p 8000:8000 quay.io/geodocker/jupyter-geopyspark:6
+
 Uninstalling
 ------------
 
@@ -241,7 +272,6 @@ To uninstall GeoPySpark, run the following in the terminal:
 
    pip uninstall geopyspark
    rm .local/bin/geopyspark
-
 
 Contributing
 ------------

--- a/README.rst
+++ b/README.rst
@@ -75,22 +75,6 @@ Where the first command installs the python code from PyPi and the second
 downloads the backend, jar file. If no path is given when downloading the jar,
 then it will be downloaded to wherever GeoPySpark was installed at.
 
-What's With That Weird Pip Install?
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-"What's with that weird pip install?", you may be asking yourself. The reason
-for its unusualness is due to how GeoPySpark functions. Because this library
-is a python binding for a Scala project, we need to be able to access the
-Scala backend. To do this, we plug into PySpark which acts as a bridge between
-Python and Scala. However, in order to achieve this the Scala code needs to be
-assembled into a jar file. This poses a problem due to its size (117.7 MB at
-v0.1.0-RC!). To get around the size constraints of PyPi, we thus utilized this
-method of distribution where the jar must be downloaded in a serperate command
-when using ``pip install``.
-
-Note:
-  Installing from source does not require the seperate download of the jar.
-
 If you would rather install from source, you can do so by running the following
 in the terminal:
 
@@ -121,6 +105,22 @@ Replace ``<your python version`` with whatever Python version
 ``virtualenvwrapper`` is set to. Installation in a virtual environment can be
 a bit weird with GeoPySpark. This is why you need to export the
 ``PYTHONPATH`` before installing to ensure that it performs correctly.
+
+What's With That Weird Pip Install?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+"What's with that weird pip install?", you may be asking yourself. The reason
+for its unusualness is due to how GeoPySpark functions. Because this library
+is a python binding for a Scala project, we need to be able to access the
+Scala backend. To do this, we plug into PySpark which acts as a bridge between
+Python and Scala. However, in order to achieve this the Scala code needs to be
+assembled into a jar file. This poses a problem due to its size (117.7 MB at
+v0.1.0-RC!). To get around the size constraints of PyPi, we thus utilized this
+method of distribution where the jar must be downloaded in a serperate command
+when using ``pip install``.
+
+Note:
+  Installing from source does not require the seperate download of the jar.
 
 Make Targets
 ^^^^^^^^^^^^

--- a/README.rst
+++ b/README.rst
@@ -114,7 +114,13 @@ enter the environment and run the following:
 
    git clone https://github.com/locationtech-labs/geopyspark.git
    cd geopyspark
+   export PYTHONPATH=$VIRTUAL_ENV/lib/<your python version>/site-packages
    make virtual-install
+
+Replace ``<your python version`` with whatever Python version
+``virtualenvwrapper`` is set to. Installation in a virtual environment can be
+a bit weird with GeoPySpark. This is why you need to export the
+``PYTHONPATH`` before installing to ensure that it performs correctly.
 
 Make Targets
 ^^^^^^^^^^^^

--- a/README.rst
+++ b/README.rst
@@ -122,6 +122,64 @@ when using ``pip install``.
 Note:
   Installing from source does not require the seperate download of the jar.
 
+Developing GeoPySpark With GeoNotebook
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`GeoNotebook <https://github.com/OpenGeoscience/geonotebook>`_ is a Jupyter
+notebook extension that specializes in working with GeoSpatial data. GeoPySpark
+can be used on this notebook; which allows for a more interactive experience
+when using the library. There two setups one could have when using these two
+tools: one for users and the other for developers. Both instances have the same
+begining steps and then diverge later. Therefore, let's take a look at what
+needs to be done first.
+
+.. code:: console
+
+   mkdir workspace
+   cd workspace
+   mkvirtualenv workspace
+
+The first thing to do is to create a new directory, ``workspace`` (or whatever
+you want to name it) and then using ``virtualenvwrapper`` to create a new
+``virtualenv`` that will contain both GeoPySpark, GeoNotebook, and their
+dependencies.
+
+Now this is where the installation for a user and developer diverge. To install
+for a user, simply follow the installation instructions for both GeoPySpark and
+GeoNotebook.
+
+For a developer, follow the instructions in the GeoNotebook README on how to
+install for development. To isntall GeoPySpark, clone the latest version from
+Github and then do this command:
+
+.. code:: console
+
+   pip install -e .
+
+This will install GeoPySpark in "editable" mode.
+
+To actaully run the notebooks, go to the folder where they are saved and run
+the command:
+
+.. code:: console
+
+   jupyter notebook
+
+This will open up the jupyter hub and will allow you to work on your notebooks.
+
+One last optional (but very recomended!) step for developers is to place these
+two lines of code at the top of your notebook.
+
+.. code:: console
+
+   %load_ext autoreload
+   $autoreload 2
+
+This will make it so that you don't have to leave the notebook for your changes
+to take affect. Rather, you just have to reimport the module and it will be
+updated.
+
+
 Make Targets
 ^^^^^^^^^^^^
 

--- a/README.rst
+++ b/README.rst
@@ -107,6 +107,15 @@ Note:
   If you have somehow altered the global behavior of ``sbt`` this install may
   not work correctly.
 
+A third option is to install GeoPySpark in a virtual environment. To do this,
+enter the environment and run the following:
+
+.. code:: console
+
+   git clone https://github.com/locationtech-labs/geopyspark.git
+   cd geopyspark
+   make virtual-install
+
 Make Targets
 ^^^^^^^^^^^^
 

